typing/value_rec_check.ml

```diff
@@ -1323,19 +1447,23 @@ let is_valid_recursive_expression idlist expr : sd option =
   match expr.exp_desc with
   | Texp_function _ ->
      (* Fast path: functions can never have invalid recursive references *)
-     Some Static
+     Some (Static Function)
   | _ ->
-     match classify_expression expr with
-     | (Static | Constant) as recursive ->
-        let ty = expression expr Return in
-        if Env.unguarded ty idlist = [] then Some recursive else None
-     | Not_recursive ->
-        let ty = expression expr Return in
-        if Env.unguarded ty idlist = [] && Env.dependent ty idlist = []
-        then Some Not_recursive
-        else None
-     | Class ->
-        assert false (* Not generated by [classify_expression] *)
+     let rkind = classify_expression expr in
+     let is_valid =
+       match rkind with
+       | Static _ | Constant ->
+         (* The expression has known size or is constant *)
+         let ty = expression expr Return in
+         Env.unguarded ty idlist = []
+       | Not_recursive ->
+         (* The expression has unknown size *)
+         let ty = expression expr Return in
+         Env.unguarded ty idlist = [] && Env.dependent ty idlist = []
+       | Class ->
+         assert false (* Not generated by [classify_expression] *)
+     in
+     if is_valid then Some rkind else None
```

Comment:
As discussed with @lthls, this separation suggests that we could in fact move the `classify_*` functions, which are much more related to compilation than type-checking, to his new Value_rec_compiler module.

Reply:
I think I will leave that for the next PR.

Reply:
(I think that this point is now moot with the split in two phases.)